stdlib/asm/collections/smt.masm

```diff
@@ -0,0 +1,201 @@
+#! input:  [v, ...]
+#! output: [v >> 48, ...]
+proc.shr_48
+    u32split
+    push.65536
+    mul
+    u32split
+    swap
+    drop
+    swap
+    drop
+end
+
+#! input:  [v, ...]
+#! output: [v >> 32, ...]
+proc.shr_32
+    u32split
+    swap
+    drop
+end
+
+#! input:  [v, ...]
+#! output: [v >> 16, ...]
+proc.shr_16
+    u32split
+    push.65536
+    mul
+    u32split
+    push.4294967296
+    mul
+    swap.2
+    push.65536
+    mul
+    u32split
+    swap
+    drop
+    add
+    add
+end
+
+#! input:  [V, K, R, ...]
+#! output: [V, R, ...]
+proc.get_16
+    # compute remaining path key `(k3 << 16) >> 16`
+    # [K', V, K, R, ...]
+    dupw.1 u32split push.65536 mul u32split drop
+    push.65536 mul u32split swap drop push.4294967296
+    mul add
+
+    # execute the permutation with the depth as domain
+    # [N, V, K, R, ...]
+    dupw.1 push.0.16.0.0 movdnw.2 hperm dropw swapw dropw
+
+    # add flag on whether value is zero
+    # [V == 0, N, V, K, R, ...]
+    dupw.1 padw eqw movdn.8 dropw dropw
+
+    # create a "zero or empty constant" word
+    # [N', V, K, R, ...]
+    dup push.17483286922353768131 mul movup.5 dup.2 not mul add
+    dup.1 push.353378057542380712 mul movup.5 dup.3 not mul add
+    dup.2 push.1935183237414585408 mul movup.5 dup.4 not mul add
+    dup.3 push.4820339620987989650 mul movup.5 dup.5 not mul add
+    movup.4 drop
+
+    # check the opening value
+    # [V', N', V, K, R, ...]
+    dupw.3 dup.12 exec.shr_48 push.16 mtree_get swapw dropw
+
+    # assert open and return values
+    # [V, R, ...]
+    assert_eqw swapw dropw
```

Comment:
minor nit here and in other places: we usually put the resulting stack state after the statements. So, for example, in other places we would write this as:
```
# assert open and return values
assert_eqw swapw dropw
# => [V, R, ...]
```

Also nit: I think the comment should say "assert the opening matches the computed node value and return".

Reply:
Fixed!